src/app/common/httpclient.ts

```diff
@@ -108,68 +95,72 @@ export class HttpClient {
-            .catch(e => {
+        req.headers.set('Access-Token', 'Bearer ' + conn.accessToken);
+        return this._http.request(req)
+            .catch((e, c) => {
                 // Status code 0 possible causes:
                 // Untrusted certificate
                 // Windows auth, prevents CORS headers from being accessed
                 // Service not responding
-                if (e.status == 0) {
-                    //
-                    // The first request to the API fails because windows auth has not started yet.
-                    // We repeat the request because in this case the next request will succeed.
-                    return this._http.request(req).toPromise()
-                        .catch(err => {
-                            //
-                            // Check to see if connected
-                            return this._http.options(this._conn.url).toPromise()
-                                .catch(e => {
-                                    this._connectSvc.reconnect();
-                                    return Promise.reject("Not connected");
+                if (e instanceof Response) {
+                    if (e.status == 0) {
+                        return this._http.request(req)
+                            .catch((err, _) => {
+                                // Check to see if connected
+                                return this._http.options(this._conn.url)
+                                    .catch((e, c2) => {
+                                        this._connectSvc.reconnect();
+                                        this.handleHttpError(err);
+                                        return c2
+                                    })
                                 })
-                                .then(r => {
-                                    return this.handleHttpError(err);
-                                });
-                        });
+                    }
+                    this.handleHttpError(e, warn);
+                    throw e
                 }
-                return this.handleHttpError(e, warn);
-            });
+                let apiError = <ApiError>({
+                    title: "unknown error",
+                    detail: JSON.stringify(e),
+                })
+                this._notificationService.apiError(apiError)
+                throw apiError
+            })
+    }
+
+    public request(url: string, options?: RequestOptionsArgs, warn?: boolean): Promise<Response> {
+        return this.requestOnConnection(url, options, warn).toPromise()
     }
 
-    private handleHttpError(err, warn?: boolean): Promise<any> {
+    private requestOnConnection(url: string, options?: RequestOptionsArgs, warn?: boolean): Observable<Response> {
+        if (this._conn) {
+            return this.performRequest(this._conn, url, options, warn)
+        } else {
+            return this.runtime.ConnectToIISHost().mergeMap(c => {
```

Comment:
please rename c with conn for readability

Reply:
renamed (next commit)